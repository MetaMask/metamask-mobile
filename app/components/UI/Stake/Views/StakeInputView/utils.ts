import {
  STAKE_INPUT_ACTION_TO_LABEL_MAP,
  STAKE_INPUT_VIEW_ACTIONS,
} from './StakeInputView.types';

export const getStakeInputViewTitle = (
  action: STAKE_INPUT_VIEW_ACTIONS,
  tokenSymbol: string,
  isEth = false,
) => {
  const prefix = STAKE_INPUT_ACTION_TO_LABEL_MAP[action];
  const suffix = isEth ? 'ETH' : tokenSymbol;

  return `${prefix} ${suffix}`;
};
