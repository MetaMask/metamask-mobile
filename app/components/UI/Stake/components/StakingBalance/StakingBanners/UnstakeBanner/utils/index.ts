import { strings } from '../../../../../../../../../locales/i18n';

/**
 * Allows for flexible rendering of unstaking banner countdown.
 *
 * Examples:
 * - Unstaking 0.0172 ETH in progress. Come back later to claim it. (days: 0, hours: 0) (default)
 * - Unstaking 0.0172 ETH in progress. Come back in 1 day to claim it. (days: 1, hours: 0).
 * - Unstaking 0.0172 ETH in progress. Come back in 1 hour to claim it. (days: 0, hours: 1).
 * - Unstaking 0.0172 ETH in progress. Come back in 2 days and 3 hours. (days: 2, hours: 3).
 */
export const renderUnstakingTimeRemaining = (
  { days, hours, minutes }: { days: number; hours: number; minutes: number },
  amountEth: string,
) => {
  if (!days && !hours && !minutes)
    return strings('stake.banner_text.unstaking_in_progress.default', {
      amountEth,
    });

  const baseCopy = strings('stake.banner_text.unstaking_in_progress.base', {
    amountEth,
  });

  const minuteString = minutes
    ? `${strings('stake.banner_text.approximately')} ${minutes} ${strings(
        'stake.minute',
        {
          count: minutes,
        },
      )}`
    : '';

  const dayString = days
    ? `${days} ${strings('stake.day', { count: days })}`
    : '';

  const hourString = hours
    ? `${hours} ${strings('stake.hour', { count: hours })}`
    : '';

  const andString =
    days && hours
      ? ` ${strings('stake.banner_text.unstaking_in_progress.and')} `
      : '';

  const toClaimString = strings(
    'stake.banner_text.unstaking_in_progress.to_claim_it',
  );

  if (!days && !hours && minutes) {
    return `${baseCopy} ${minuteString} ${toClaimString}`.trim();
  }

  return `${baseCopy} ${dayString}${andString}${hourString} ${toClaimString}`.trim();
};
