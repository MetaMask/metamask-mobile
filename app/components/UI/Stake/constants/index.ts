/* eslint-disable import/prefer-default-export */
export const isPooledStakingFeatureEnabled = () =>
  process.env.MM_POOLED_STAKING_UI_ENABLED === 'true';

export const POOLED_STAKING_FAQ_URL =
  'https://support.metamask.io/metamask-portfolio/move-crypto/stake/staking-pool/';
