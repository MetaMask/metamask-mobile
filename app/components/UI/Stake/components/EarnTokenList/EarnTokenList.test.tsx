/* eslint-disable import/no-namespace */
import React from 'react';
import EarnTokenList from '.';
import renderWithProvider from '../../../../../util/test/renderWithProvider';
import { MOCK_ACCOUNTS_CONTROLLER_STATE } from '../../../../../util/test/accountsControllerTestUtils';
import initialRootState from '../../../../../util/test/initial-root-state';
import { Metrics, SafeAreaProvider } from 'react-native-safe-area-context';
import { strings } from '../../../../../../locales/i18n';
import {
  MOCK_SUPPORTED_EARN_TOKENS_NO_FIAT_BALANCE,
  MOCK_USDC_BASE_MAINNET_ASSET,
} from '../../__mocks__/mockData';
import Engine from '../../../../../core/Engine';
import * as tokenUtils from '../../utils/token';
import * as useStakingEligibilityHook from '../../hooks/useStakingEligibility';
import * as stakeConstants from '../../constants';
import * as portfolioNetworkUtils from '../../../../../util/networks';
import { act, fireEvent } from '@testing-library/react-native';

jest.mock('../../../../../core/Engine', () => ({
  context: {
    NetworkController: {
      getNetworkClientById: () => ({
        configuration: {
          chainId: '0x1',
          rpcUrl: 'https://mainnet.infura.io/v3',
          ticker: 'ETH',
          type: 'custom',
        },
      }),
      findNetworkClientIdByChainId: () => 'mainnet',
      setActiveNetwork: jest.fn(),
    },
  },
}));

jest.mock('../../../../../util/networks', () => ({
  isPortfolioViewEnabled: jest.fn().mockReturnValue(true),
}));

jest.mock('../../constants', () => ({
  isStablecoinLendingFeatureEnabled: jest.fn().mockReturnValue(true),
}));

const mockNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
    }),
  };
});

jest.mock('../../../../../util/networks', () => ({
  ...jest.requireActual('../../../../../util/networks'),
  getNetworkImageSource: jest.fn().mockReturnValue(10),
}));

const initialState = {
  ...initialRootState,
  engine: {
    ...initialRootState.engine,
    backgroundState: {
      ...initialRootState.engine.backgroundState,
      AccountsController: MOCK_ACCOUNTS_CONTROLLER_STATE,
    },
  },
};

const initialMetrics: Metrics = {
  frame: { x: 0, y: 0, width: 320, height: 640 },
  insets: { top: 0, left: 0, right: 0, bottom: 0 },
};

let useStakingEligibilitySpy: jest.SpyInstance;
let getSupportedEarnTokensSpy: jest.SpyInstance;
let filterEligibleTokensSpy: jest.SpyInstance;

describe('EarnTokenList', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    useStakingEligibilitySpy = jest
      .spyOn(useStakingEligibilityHook, 'default')
      .mockReturnValue({
        isEligible: true,
        isLoadingEligibility: false,
        refreshPooledStakingEligibility: jest.fn().mockResolvedValue({
          isEligible: true,
        }),
        error: '',
      });

    getSupportedEarnTokensSpy = jest
      .spyOn(tokenUtils, 'getSupportedEarnTokens')
      .mockReturnValue(MOCK_SUPPORTED_EARN_TOKENS_NO_FIAT_BALANCE);

    filterEligibleTokensSpy = jest.spyOn(tokenUtils, 'filterEligibleTokens');
  });

  it('render matches snapshot', () => {
    const { toJSON, getByText, getAllByText } = renderWithProvider(
      <SafeAreaProvider initialMetrics={initialMetrics}>
        <EarnTokenList />
      </SafeAreaProvider>,
      {
        state: initialState,
      },
    );

    expect(toJSON()).toMatchSnapshot();

    // Bottom Sheet Title
    expect(getByText(strings('stake.select_a_token'))).toBeDefined();

    // Upsell Banner
    expect(getByText(strings('stake.you_could_earn'))).toBeDefined();
    expect(getByText(strings('stake.per_year_on_your_tokens'))).toBeDefined();

    // Token List
    // Ethereum
    expect(getAllByText('Ethereum').length).toBe(1);
    expect(getAllByText('2.3% APR').length).toBe(1);

    // DAI
    expect(getByText('Dai Stablecoin')).toBeDefined();
    expect(getByText('5.0% APR')).toBeDefined();

    // USDT
    expect(getByText('Tether USD')).toBeDefined();
    expect(getByText('4.1% APR')).toBeDefined();

    // USDC
    expect(getByText('USDC')).toBeDefined();
    expect(getAllByText('4.5% APR').length).toBe(1);

    expect(getSupportedEarnTokensSpy).toHaveBeenCalled();
    expect(filterEligibleTokensSpy).toHaveBeenCalled();
  });

  it('does not render the EarnTokenList when required feature flags are disabled', () => {
    jest
      .spyOn(stakeConstants, 'isStablecoinLendingFeatureEnabled')
      .mockReturnValueOnce(false);
    jest
      .spyOn(portfolioNetworkUtils, 'isPortfolioViewEnabled')
      .mockReturnValueOnce(false);

    const { toJSON } = renderWithProvider(<EarnTokenList />);

    expect(toJSON()).toBeNull();
  });

  it('changes active network if selected token is on a different network', async () => {
    getSupportedEarnTokensSpy = jest
      .spyOn(tokenUtils, 'getSupportedEarnTokens')
      .mockReturnValue([MOCK_USDC_BASE_MAINNET_ASSET]);

    const { getByText } = renderWithProvider(
      <SafeAreaProvider initialMetrics={initialMetrics}>
        <EarnTokenList />
      </SafeAreaProvider>,
      {
        state: initialState,
      },
    );

    const baseUsdc = getByText('USDC');

    await act(() => {
      fireEvent.press(baseUsdc);
    });

    expect(
      Engine.context.NetworkController.setActiveNetwork,
    ).toHaveBeenCalledWith('mainnet');

    expect(getSupportedEarnTokensSpy).toHaveBeenCalled();
    expect(filterEligibleTokensSpy).toHaveBeenCalled();
  });

  it('hides staking tokens if user is not eligible', () => {
    useStakingEligibilitySpy.mockReturnValue({
      isEligible: false,
      isLoadingEligibility: false,
      refreshPooledStakingEligibility: jest.fn().mockResolvedValue({
        isEligible: false,
      }),
      error: '',
    });

    const { queryByText, getByText } = renderWithProvider(
      <SafeAreaProvider initialMetrics={initialMetrics}>
        <EarnTokenList />
      </SafeAreaProvider>,
      {
        state: initialState,
      },
    );

    expect(queryByText('Ethereum')).toBeNull();
    expect(queryByText('Staked Ethereum')).toBeNull();

    expect(getByText('Dai Stablecoin')).toBeDefined();
    expect(getByText('USDC')).toBeDefined();
    expect(getByText('Tether USD')).toBeDefined();

    expect(getSupportedEarnTokensSpy).toHaveBeenCalled();

    expect(filterEligibleTokensSpy).toHaveBeenCalledWith(
      MOCK_SUPPORTED_EARN_TOKENS_NO_FIAT_BALANCE,
      { canStake: false, canLend: true },
    );
  });

  it('hides lending tokens if user is not eligible', () => {
    filterEligibleTokensSpy.mockImplementationOnce(() =>
      tokenUtils.filterEligibleTokens(
        MOCK_SUPPORTED_EARN_TOKENS_NO_FIAT_BALANCE,
        { canStake: true, canLend: false },
      ),
    );

    const { queryByText, getAllByText } = renderWithProvider(
      <SafeAreaProvider initialMetrics={initialMetrics}>
        <EarnTokenList />
      </SafeAreaProvider>,
      {
        state: initialState,
      },
    );

    expect(getAllByText('Ethereum').length).toBe(1);
    expect(queryByText('Staked Ethereum')).toBeDefined();

    expect(queryByText('Dai Stablecoin')).toBeNull();
    expect(queryByText('USDC')).toBeNull();
    expect(queryByText('Tether USD')).toBeNull();
    expect(queryByText('USD Coin')).toBeNull();

    expect(getSupportedEarnTokensSpy).toHaveBeenCalled();
    expect(filterEligibleTokensSpy).toHaveBeenCalled();
  });

  it('redirects to StakeInputView with pooled staking navigation params for staking token', async () => {
    const { getByText } = renderWithProvider(
      <SafeAreaProvider initialMetrics={initialMetrics}>
        <EarnTokenList />
      </SafeAreaProvider>,
      {
        state: initialState,
      },
    );

    const ethButton = getByText('Ethereum');

    await act(() => {
      fireEvent.press(ethButton);
    });

    expect(mockNavigate).toHaveBeenCalledWith('StakeScreens', {
      params: {
        action: 'STAKE',
        token: {
          address: '0xabc',
          aggregators: [],
          balance: '',
          balanceFiat: '',
          chainId: '0x1',
          decimals: 18,
          image: '',
          isETH: true,
          isNative: true,
          isStaked: false,
          logo: '',
          name: 'Ethereum',
          symbol: 'Ethereum',
          ticker: 'ETH',
          tokenBalanceFormatted: ' ETH',
        },
      },
      screen: 'Stake',
    });
  });

  it('redirect to StakeInputView with stablecoin lending navigation params for lending token', async () => {
    const { getByText } = renderWithProvider(
      <SafeAreaProvider initialMetrics={initialMetrics}>
        <EarnTokenList />
      </SafeAreaProvider>,
      {
        state: initialState,
      },
    );

    const usdcButton = getByText('USDC');

    await act(() => {
      fireEvent.press(usdcButton);
    });

    expect(mockNavigate).toHaveBeenCalledWith('StakeScreens', {
      params: {
        action: 'LEND',
        token: {
          address: '0xabc',
          aggregators: [],
          balance: '',
          balanceFiat: 'tokenBalanceLoading',
          chainId: '0x1',
          decimals: 6,
          image: '',
          isETH: false,
          isNative: false,
          isStaked: false,
          logo: '',
          name: 'USDC',
          symbol: 'USDC',
          ticker: 'USDC',
          tokenBalanceFormatted: 'tokenBalanceLoading',
        },
      },
      screen: 'Stake',
    });
  });
});
