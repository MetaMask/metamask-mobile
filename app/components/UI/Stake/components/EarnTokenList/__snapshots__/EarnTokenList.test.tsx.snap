// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EarnTokenList render matches snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "bottom": 0,
          "justifyContent": "flex-end",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#00000099",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          {
            "opacity": 0,
          },
        ]
      }
    >
      <TouchableOpacity
        onPress={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        collapsable={false}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "backgroundColor": "#ffffff",
              "borderColor": "#BBC0C566",
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
              "borderWidth": 1,
              "maxHeight": 1334,
              "overflow": "hidden",
              "paddingBottom": 0,
              "shadowColor": "#0000001A",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 1,
              "shadowRadius": 40,
            },
            {
              "transform": [
                {
                  "translateY": 1334,
                },
              ],
            },
          ]
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "alignSelf": "stretch",
              "padding": 4,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "#BBC0C566",
                "borderRadius": 2,
                "height": 4,
                "width": 40,
              }
            }
          />
        </View>
        <View
          style={
            [
              {
                "backgroundColor": "#ffffff",
                "flexDirection": "row",
                "padding": 16,
              },
              false,
            ]
          }
          testID="header"
        >
          <View
            style={
              {
                "width": undefined,
              }
            }
          >
            <View
              onLayout={[Function]}
            />
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "marginHorizontal": 16,
              }
            }
          >
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#141618",
                  "fontFamily": "EuclidCircularB-Bold",
                  "fontSize": 16,
                  "fontWeight": "700",
                  "letterSpacing": 0,
                  "lineHeight": 24,
                }
              }
            >
              Select a token
            </Text>
          </View>
          <View
            style={
              {
                "width": undefined,
              }
            }
          >
            <View
              onLayout={[Function]}
            />
          </View>
        </View>
        <RCTScrollView
          collapsable={false}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          style={
            {
              "gap": 12,
              "maxHeight": 552,
              "paddingHorizontal": 16,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#f2f4f6",
                  "borderRadius": 8,
                  "gap": 8,
                  "paddingHorizontal": 16,
                  "paddingVertical": 24,
                }
              }
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#141618",
                    "fontFamily": "EuclidCircularB-Bold",
                    "fontSize": 18,
                    "fontWeight": "700",
                    "letterSpacing": 0,
                    "lineHeight": 24,
                  }
                }
              >
                You could earn
              </Text>
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#1c8234",
                    "fontFamily": "EuclidCircularB-Bold",
                    "fontSize": 32,
                    "fontWeight": "700",
                    "letterSpacing": 0,
                    "lineHeight": 40,
                  }
                }
              >
                $454
              </Text>
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#6a737d",
                    "fontFamily": "EuclidCircularB-Regular",
                    "fontSize": 14,
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 22,
                  }
                }
              >
                per year on your tokens
              </Text>
            </View>
            <View
              style={
                {
                  "padding": 16,
                }
              }
            >
              <TouchableOpacity
                onPress={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "gap": 16,
                    }
                  }
                >
                  <View
                    onLayout={[Function]}
                    style={
                      {
                        "alignSelf": "flex-start",
                        "position": "relative",
                      }
                    }
                    testID="badge-wrapper-badge"
                  >
                    <View>
                      <View
                        useNativeDriver={true}
                      >
                        <Image
                          fadeIn={true}
                          onError={[Function]}
                          onLoadEnd={[Function]}
                          source={
                            {
                              "uri": "",
                            }
                          }
                          style={
                            [
                              {
                                "borderRadius": 12,
                                "height": 24,
                                "width": 24,
                              },
                              undefined,
                              false,
                              false,
                              {
                                "flexShrink": 0,
                                "height": 32,
                                "width": 32,
                              },
                            ]
                          }
                          testID="earn-token-list-item-Ethereum-0x1"
                        />
                        <View
                          collapsable={false}
                          style={
                            {
                              "bottom": 0,
                              "left": 0,
                              "opacity": 1,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                            }
                          }
                        >
                          <View
                            style={
                              [
                                {
                                  "borderRadius": 12,
                                  "flexShrink": 0,
                                  "height": 32,
                                  "width": 32,
                                },
                                {
                                  "backgroundColor": "#eee",
                                },
                                undefined,
                              ]
                            }
                          />
                        </View>
                      </View>
                    </View>
                    <View
                      style={
                        {
                          "alignItems": "center",
                          "aspectRatio": 1,
                          "height": 0,
                          "justifyContent": "center",
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                          "transform": [
                            {
                              "translateX": 0,
                            },
                            {
                              "translateY": -0,
                            },
                          ],
                        }
                      }
                    >
                      <View
                        onLayout={[Function]}
                        style={
                          {
                            "alignItems": "center",
                            "aspectRatio": 1,
                            "height": "50%",
                            "justifyContent": "center",
                            "maxHeight": 24,
                            "minHeight": 8,
                            "opacity": 0,
                          }
                        }
                        testID="badgenetwork"
                      >
                        <View
                          style={
                            {
                              "alignItems": "center",
                              "backgroundColor": "#ffffff",
                              "borderColor": "#ffffff",
                              "borderRadius": 8,
                              "borderWidth": 2,
                              "height": 32,
                              "justifyContent": "center",
                              "overflow": "hidden",
                              "shadowColor": "#0000001A",
                              "shadowOffset": {
                                "height": 2,
                                "width": 0,
                              },
                              "shadowOpacity": 1,
                              "shadowRadius": 4,
                              "transform": [
                                {
                                  "scale": 1,
                                },
                              ],
                              "width": 32,
                            }
                          }
                        >
                          <Image
                            onError={[Function]}
                            resizeMode="contain"
                            source={10}
                            style={
                              {
                                "height": 32,
                                "width": 32,
                              }
                            }
                            testID="network-avatar-image"
                          />
                        </View>
                      </View>
                    </View>
                  </View>
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#141618",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                      }
                    }
                  >
                    Ethereum
                  </Text>
                </View>
                <View
                  style={
                    {
                      "alignItems": "flex-end",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#1c8234",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                      }
                    }
                  >
                    2.3% APR
                  </Text>
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#6a737d",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 12,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 20,
                      }
                    }
                  >
                     ETH
                  </Text>
                </View>
              </TouchableOpacity>
            </View>
            <View
              style={
                {
                  "padding": 16,
                }
              }
            >
              <TouchableOpacity
                onPress={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "gap": 16,
                    }
                  }
                >
                  <View
                    onLayout={[Function]}
                    style={
                      {
                        "alignSelf": "flex-start",
                        "position": "relative",
                      }
                    }
                    testID="badge-wrapper-badge"
                  >
                    <View>
                      <View
                        style={
                          {
                            "backgroundColor": "#ffffff",
                            "borderRadius": 16,
                            "flexShrink": 0,
                            "height": 32,
                            "overflow": "hidden",
                            "width": 32,
                          }
                        }
                      >
                        <Image
                          onError={[Function]}
                          resizeMode="contain"
                          source={
                            {
                              "uri": "",
                            }
                          }
                          style={
                            {
                              "flex": 1,
                              "height": undefined,
                              "width": undefined,
                            }
                          }
                          testID="token-avatar-image"
                        />
                      </View>
                    </View>
                    <View
                      style={
                        {
                          "alignItems": "center",
                          "aspectRatio": 1,
                          "height": 0,
                          "justifyContent": "center",
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                          "transform": [
                            {
                              "translateX": 0,
                            },
                            {
                              "translateY": -0,
                            },
                          ],
                        }
                      }
                    >
                      <View
                        onLayout={[Function]}
                        style={
                          {
                            "alignItems": "center",
                            "aspectRatio": 1,
                            "height": "50%",
                            "justifyContent": "center",
                            "maxHeight": 24,
                            "minHeight": 8,
                            "opacity": 0,
                          }
                        }
                        testID="badgenetwork"
                      >
                        <View
                          style={
                            {
                              "alignItems": "center",
                              "backgroundColor": "#ffffff",
                              "borderColor": "#ffffff",
                              "borderRadius": 8,
                              "borderWidth": 2,
                              "height": 32,
                              "justifyContent": "center",
                              "overflow": "hidden",
                              "shadowColor": "#0000001A",
                              "shadowOffset": {
                                "height": 2,
                                "width": 0,
                              },
                              "shadowOpacity": 1,
                              "shadowRadius": 4,
                              "transform": [
                                {
                                  "scale": 1,
                                },
                              ],
                              "width": 32,
                            }
                          }
                        >
                          <Image
                            onError={[Function]}
                            resizeMode="contain"
                            source={10}
                            style={
                              {
                                "height": 32,
                                "width": 32,
                              }
                            }
                            testID="network-avatar-image"
                          />
                        </View>
                      </View>
                    </View>
                  </View>
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#141618",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                      }
                    }
                  >
                    Dai Stablecoin
                  </Text>
                </View>
                <View
                  style={
                    {
                      "alignItems": "flex-end",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#1c8234",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                      }
                    }
                  >
                    5.0% APR
                  </Text>
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#6a737d",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 12,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 20,
                      }
                    }
                  >
                    tokenBalanceLoading
                  </Text>
                </View>
              </TouchableOpacity>
            </View>
            <View
              style={
                {
                  "padding": 16,
                }
              }
            >
              <TouchableOpacity
                onPress={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "gap": 16,
                    }
                  }
                >
                  <View
                    onLayout={[Function]}
                    style={
                      {
                        "alignSelf": "flex-start",
                        "position": "relative",
                      }
                    }
                    testID="badge-wrapper-badge"
                  >
                    <View>
                      <View
                        style={
                          {
                            "backgroundColor": "#ffffff",
                            "borderRadius": 16,
                            "flexShrink": 0,
                            "height": 32,
                            "overflow": "hidden",
                            "width": 32,
                          }
                        }
                      >
                        <Image
                          onError={[Function]}
                          resizeMode="contain"
                          source={
                            {
                              "uri": "",
                            }
                          }
                          style={
                            {
                              "flex": 1,
                              "height": undefined,
                              "width": undefined,
                            }
                          }
                          testID="token-avatar-image"
                        />
                      </View>
                    </View>
                    <View
                      style={
                        {
                          "alignItems": "center",
                          "aspectRatio": 1,
                          "height": 0,
                          "justifyContent": "center",
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                          "transform": [
                            {
                              "translateX": 0,
                            },
                            {
                              "translateY": -0,
                            },
                          ],
                        }
                      }
                    >
                      <View
                        onLayout={[Function]}
                        style={
                          {
                            "alignItems": "center",
                            "aspectRatio": 1,
                            "height": "50%",
                            "justifyContent": "center",
                            "maxHeight": 24,
                            "minHeight": 8,
                            "opacity": 0,
                          }
                        }
                        testID="badgenetwork"
                      >
                        <View
                          style={
                            {
                              "alignItems": "center",
                              "backgroundColor": "#ffffff",
                              "borderColor": "#ffffff",
                              "borderRadius": 8,
                              "borderWidth": 2,
                              "height": 32,
                              "justifyContent": "center",
                              "overflow": "hidden",
                              "shadowColor": "#0000001A",
                              "shadowOffset": {
                                "height": 2,
                                "width": 0,
                              },
                              "shadowOpacity": 1,
                              "shadowRadius": 4,
                              "transform": [
                                {
                                  "scale": 1,
                                },
                              ],
                              "width": 32,
                            }
                          }
                        >
                          <Image
                            onError={[Function]}
                            resizeMode="contain"
                            source={10}
                            style={
                              {
                                "height": 32,
                                "width": 32,
                              }
                            }
                            testID="network-avatar-image"
                          />
                        </View>
                      </View>
                    </View>
                  </View>
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#141618",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                      }
                    }
                  >
                    USDC
                  </Text>
                </View>
                <View
                  style={
                    {
                      "alignItems": "flex-end",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#1c8234",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                      }
                    }
                  >
                    4.5% APR
                  </Text>
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#6a737d",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 12,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 20,
                      }
                    }
                  >
                    tokenBalanceLoading
                  </Text>
                </View>
              </TouchableOpacity>
            </View>
            <View
              style={
                {
                  "padding": 16,
                }
              }
            >
              <TouchableOpacity
                onPress={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "gap": 16,
                    }
                  }
                >
                  <View
                    onLayout={[Function]}
                    style={
                      {
                        "alignSelf": "flex-start",
                        "position": "relative",
                      }
                    }
                    testID="badge-wrapper-badge"
                  >
                    <View>
                      <View
                        style={
                          {
                            "backgroundColor": "#ffffff",
                            "borderRadius": 16,
                            "flexShrink": 0,
                            "height": 32,
                            "overflow": "hidden",
                            "width": 32,
                          }
                        }
                      >
                        <Image
                          onError={[Function]}
                          resizeMode="contain"
                          source={
                            {
                              "uri": "",
                            }
                          }
                          style={
                            {
                              "flex": 1,
                              "height": undefined,
                              "width": undefined,
                            }
                          }
                          testID="token-avatar-image"
                        />
                      </View>
                    </View>
                    <View
                      style={
                        {
                          "alignItems": "center",
                          "aspectRatio": 1,
                          "height": 0,
                          "justifyContent": "center",
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                          "transform": [
                            {
                              "translateX": 0,
                            },
                            {
                              "translateY": -0,
                            },
                          ],
                        }
                      }
                    >
                      <View
                        onLayout={[Function]}
                        style={
                          {
                            "alignItems": "center",
                            "aspectRatio": 1,
                            "height": "50%",
                            "justifyContent": "center",
                            "maxHeight": 24,
                            "minHeight": 8,
                            "opacity": 0,
                          }
                        }
                        testID="badgenetwork"
                      >
                        <View
                          style={
                            {
                              "alignItems": "center",
                              "backgroundColor": "#ffffff",
                              "borderColor": "#ffffff",
                              "borderRadius": 8,
                              "borderWidth": 2,
                              "height": 32,
                              "justifyContent": "center",
                              "overflow": "hidden",
                              "shadowColor": "#0000001A",
                              "shadowOffset": {
                                "height": 2,
                                "width": 0,
                              },
                              "shadowOpacity": 1,
                              "shadowRadius": 4,
                              "transform": [
                                {
                                  "scale": 1,
                                },
                              ],
                              "width": 32,
                            }
                          }
                        >
                          <Image
                            onError={[Function]}
                            resizeMode="contain"
                            source={10}
                            style={
                              {
                                "height": 32,
                                "width": 32,
                              }
                            }
                            testID="network-avatar-image"
                          />
                        </View>
                      </View>
                    </View>
                  </View>
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#141618",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                      }
                    }
                  >
                    Tether USD
                  </Text>
                </View>
                <View
                  style={
                    {
                      "alignItems": "flex-end",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#1c8234",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                      }
                    }
                  >
                    4.1% APR
                  </Text>
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#6a737d",
                        "fontFamily": "EuclidCircularB-Medium",
                        "fontSize": 12,
                        "fontWeight": "500",
                        "letterSpacing": 0,
                        "lineHeight": 20,
                      }
                    }
                  >
                    tokenBalanceLoading
                  </Text>
                </View>
              </TouchableOpacity>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
